src/string_object.cpp

```diff
@@ -344,6 +344,40 @@ Value StringObject::ord(Env *env) {
     return Value::integer(code);
 }
 
+Value StringObject::prepend(Env *env, size_t argc, Value *args) {
+    assert_not_frozen(env);
+
+    StringObject *original = new StringObject(*this);
+
+    auto to_str = "to_str"_s;
+    String appendable;
+    for (size_t i = 0; i < argc; i++) {
+        auto arg = args[i];
+
+        if (arg == this)
+            arg = original;
+
+        StringObject *str_obj;
+        if (arg->is_string()) {
+            str_obj = arg->as_string();
+        } else if (arg->is_integer() && arg->as_integer()->to_nat_int_t() < 0) {
+            env->raise("RangeError", "less than 0");
+        } else if (arg->is_integer()) {
+            str_obj = arg.send(env, "chr"_s)->as_string();
+        } else if (arg->respond_to(env, to_str)) {
+            str_obj = arg.send(env, to_str)->as_string();
+        } else {
+            env->raise("TypeError", "cannot call to_str", arg->inspect_str(env));
+        }
+
+        str_obj->assert_type(env, Object::Type::String, "String");
+        appendable.append(str_obj->c_str());
+    }
+    m_string.prepend(appendable.c_str());
```

Comment:
One thing that @FnControlOption worked on was making our `String` class properly handle internal null characters `\0`. I think we need to avoid dropping down to cstrs as much as possible for that reason.

```ruby
s = "foo"
s.prepend("bar\0baz")
p s # => "bar\u0000bazfoo"
```

Reply:
You should be able to add a `String::prepend()` that accepts another String....